kernel/work.c

```diff
@@ -0,0 +1,1030 @@
+ * the queue it was submitted to.  That may or may not be the queue provided
+ * on input.
+ *
+ * @retval 0 if already queued
+ * @retval 1 if unqueued and not running
+ * @retval 2 if unqueued and running
+ * @retval -EBUSY if canceling or submission was rejected by queue
+ * @retval -EINVAL if no queue is provided
+ * @retval -ENODEV if the queue is not started
+ */
+static int submit_to_queue_locked(struct k_work *work,
+				  struct k_work_q **queuep)
+{
+	int ret = 0;
+
+	if (flag_test(&work->flags, K_WORK_CANCELING_BIT)) {
+		/* Disallowed */
+		ret = -EBUSY;
+	} else if (!flag_test(&work->flags, K_WORK_QUEUED_BIT)) {
+		/* Not currently queued */
+		ret = 1;
+
+		/* If no queue specified resubmit to last queue.
+		 */
+		if (*queuep == NULL) {
+			*queuep = work->queue;
+		}
+
+		/* If the work is currently running we have to use the
+		 * queue it's running on to prevent handler
+		 * re-entrancy.
+		 */
+		if (flag_test(&work->flags, K_WORK_RUNNING_BIT)) {
+			__ASSERT_NO_MSG(work->queue != NULL);
+			*queuep = work->queue;
+			ret = 2;
+		}
+
+		int rc = queue_submit_locked(*queuep, work);
+
+		if (rc < 0) {
+			ret = rc;
+		} else {
+			flag_set(&work->flags, K_WORK_QUEUED_BIT);
+			work->queue = *queuep;
+		}
+	} else {
+		/* Already queued, do nothing. */
+	}
+
+	if (ret <= 0) {
+		*queuep = NULL;
+	}
+
+	return ret;
+}
+
+int k_work_submit_to_queue(struct k_work_q *queue,
+			    struct k_work *work)
+{
+	__ASSERT_NO_MSG(work != NULL);
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	int ret = submit_to_queue_locked(work, &queue);
+
+	k_spin_unlock(&lock, key);
+
+	/* If we changed the queue contents (as indicated by a positive ret)
+	 * the queue thread may now be ready, but we missed the reschedule
+	 * point because the lock was held.  If this is being invoked by a
+	 * preemptible thread then yield.
+	 */
+	if ((ret > 0) && (k_is_preempt_thread() != 0)) {
+		k_yield();
+	}
+
+
+	return ret;
+}
+
+/* Flush the work item if necessary.
+ *
+ * Flushing is necessary only if the work is either queued or running.
+ *
+ * Invoked with work lock held by key.
+ * Sleeps.
+ *
+ * @param work the work item that is to be flushed
+ * @param flusher state used to synchronize the flush
+ * @param key the key under which the work lock is held.
+ *
+ * @retval true if work is queued or running.  If this happens the
+ * caller must take the flusher semaphore after releasing the lock.
+ *
+ * @retval false otherwise.  No wait required.
+ */
+static bool work_flush_locked(struct k_work *work,
+			      struct z_work_flusher *flusher)
+{
+	struct k_work_q *queue = NULL;
+	bool need_flush = (flags_get(&work->flags)
+			   & (K_WORK_QUEUED | K_WORK_RUNNING)) != 0;
+
+	if (need_flush) {
+		queue = work->queue;
+		__ASSERT_NO_MSG(queue != NULL);
+
+		queue_flusher_locked(queue, work, flusher);
+		notify_queue_locked(queue);
+	}
+
+	return need_flush;
+}
+
+bool k_work_flush(struct k_work *work,
+		  struct k_work_sync *sync)
```

Comment:
(probably I am missing something here...)

Could you explain to me why an external sync struct is needed? Is it possible to declare it inside this function so it exists in the stack?

Reply:
It cannot be declared in this function, because it contains a node that is linked into the work queue to be processed when the requested work item completes, which may not happen until after this function returns.

It can be defined on the stack of the caller if the target platform is cache-coherent, but on some platforms it would have to be in memory accessible by other processors.

Reply:
Originally the motivation for `k_work_sync` was userspace access, but that's no longer relevant.  The remaining concern here is with architectures that are not cache coherent, which I understand to mean there are problems with visibility of thread stacks between processors.  We need at minimum a semaphore, and for cancel an slist node that links the semaphore into the canceling list.

If kernel folks like @dcpleung and @andyross confirm that putting such data in the thread stack won't cause problems on any architecture Zephyr needs to support then I'll refactor this.

I'll wait until this is resolved before finalizing updates.

Reply:
Sorry, @carlescufi had to call me over.

The cache coherence is specific to Xtensa SMP platforms with their L1 cache enabled, which right now means the various intel_adsp devices.  The framework we decided on for managing that is that all non-stack data is by default coherent and uncached, and thus safe (there is an __incoherent linkage specifier for apps that know what they are doing).  But stacks are cached, and it's never legal to inspect another thread's stack (strictly: it's never legal to inspect the stack of a thread that has last run on another CPU).

So yeah, that means no IPC data like semaphores or or shared lists on the stack.  Where I've had to work around this, I've relied on static data mostly, or in a pinch by stealing data in the thread struct.  If it's absolutely required, we actually could write a thin layer that would allow putting manually-flushed data on the stack.  You'd have to pack it into an exclusive cache line (Xtensa has 64 bit cache lines) and carefully invalidate and flush it as needed (maybe in combination with a spinlock acquire/release?).   Cache management gets mindbending though....

Reply:
OK, so that says we can't generally put the synchronization objects on the stack: they have to be provided by the caller and be in non-stack memory if the operation is going to work on Xtensa SMP.

The problem is that even though `k_work_sync` is documented as do-not-stack-allocate when `CONFIG_KERNEL_COHERENCE=y` nobody's going to pay any attention to that, so generic code where the caller allocates the thing on the stack is still going to break on Xtensa SMP.  So what we have now may not work right, even though there's just such a case in `test_smp_running_cancel()` in `tests/kernel/qorkq/work` which hasn't shown failures.  Maybe that's due to `arch_cohere_stacks()` flushing the cache on context switch, so it just works as long as the processor running the work queue thread context switches?

There was a similar issue with k_poll addressed in dadc6643e45cde165d474874632674c2f0f185c2 so we could instead put a `k_work_sync` object in a union with the `struct _poller` in the thread memory.

At this point unless `arch_cohere_stacks()` makes this irrelevant I'm inclined to try that.  @andyross do you see a problem with it?  Including the increase in per-thread memory requirements (`struct z_work_flusher` is rather bigger than `struct _poller`.)

Reply:
FWIW: if you want to enforce a rule like that, I don't see it as too onerous.  KERNEL_COHERENCE is a separate mode for a reason; most apps aren't going to want to follow those rules.  You can check pointers at runtime with `arch_mem_coherent()` and assert on them if you need to catch mistakes.

The problem then is that you need to make sure that any test code we have in-tree uses only coherent data.

Reply:
I've gone with the document-and-assert approach, so the API will work in the general case.